File: frontend/apps/ios/Pantopus/Features/Businesses/Team/BusinessTeamViewModel.swift
// Owner-side business team and roles roster. Members are sourced from
// businessIam, pending invites from businessSeats, and action gating from /me.

import Foundation
import Observation
import SwiftUI

// MARK: - Projected view models

/// One member row in a role section.
public struct BusinessTeamMemberRow: Identifiable, Sendable, Equatable {
    public let id: String
    public let userId: String
    public let name: String
    public let email: String?
    public let role: BusinessRole
    public let avatarGradient: GradientPair
    public let joinedText: String?
    /// True when the viewer can change/remove this member (gated on
    /// `team.manage` and never true for the owner row).
    public let canManage: Bool
}

/// One role group section (owner → viewer order).
public struct BusinessTeamSection: Identifiable, Sendable, Equatable {
    public let id: String
    public let role: BusinessRole
    public let rows: [BusinessTeamMemberRow]

    public var headerTitle: String {
        role.pluralLabel
    }
}

/// One pending seat-invite row.
public struct BusinessTeamPendingRow: Identifiable, Sendable, Equatable {
    public let id: String
    public let seatId: String
    public let name: String
    public let email: String?
    public let role: BusinessRole
    public let invitedText: String?
    public let canManage: Bool
}

/// Loaded payload for the Team screen.
public struct BusinessTeamContent: Sendable, Equatable {
    public let sections: [BusinessTeamSection]
    public let pending: [BusinessTeamPendingRow]
    public let canManage: Bool
    public let canInvite: Bool
}

/// Render state for the Team screen.
public enum BusinessTeamState: Sendable, Equatable {
    case loading
    case loaded(BusinessTeamContent)
    case empty(canInvite: Bool)
    case error(message: String)
}

// MARK: - View model

@Observable
@MainActor
public final class BusinessTeamViewModel {
    // MARK: Public state

    public private(set) var state: BusinessTeamState = .loading

    /// Assignable role presets for the change-role picker. Populated by
    /// `load()`; empty until then.
    public private(set) var rolePresets: [BusinessRolePresetDTO] = []

    // MARK: Dependencies

    private let businessId: String
    private let api: APIClient
    private let now: @Sendable () -> Date
    private let calendar: Calendar
    private let timeZone: TimeZone

    // MARK: Raw state

    private var access: BusinessTeamAccessDTO?
    private var members: [BusinessTeamMemberDTO] = []
    private var pendingSeats: [BusinessSeatDTO] = []
    private var loadedOnce = false

    init(
        businessId: String,
        api: APIClient = .shared,
        now: @escaping @Sendable () -> Date = { Date() },
        calendar: Calendar = .current,
        timeZone: TimeZone = .current,
        seed: Seed? = nil
    ) {
        self.businessId = businessId
        self.api = api
        self.now = now
        self.calendar = calendar
        self.timeZone = timeZone
        if let seed {
            access = seed.access
            members = seed.members
            pendingSeats = seed.pendingSeats
            rolePresets = seed.presets
            loadedOnce = true
            applyState()
        }
    }

    /// Preview/test seed bundle so SwiftUI previews and unit tests can
    /// render the loaded state without a network round-trip.
    public struct Seed: Sendable {
        public let access: BusinessTeamAccessDTO
        public let members: [BusinessTeamMemberDTO]
        public let pendingSeats: [BusinessSeatDTO]
        public let presets: [BusinessRolePresetDTO]

        public init(
            access: BusinessTeamAccessDTO,
            members: [BusinessTeamMemberDTO],
            pendingSeats: [BusinessSeatDTO],
            presets: [BusinessRolePresetDTO]
        ) {
            self.access = access
            self.members = members
            self.pendingSeats = pendingSeats
            self.presets = presets
        }

        /// Owner-as-viewer fixture mirroring `BusinessTeamSampleData`.
        public static let sample = Seed(
            access: BusinessTeamSampleData.access,
            members: BusinessTeamSampleData.members,
            pendingSeats: BusinessTeamSampleData.pendingSeats,
            presets: BusinessTeamSampleData.presets
        )
    }

    // MARK: - Load

    public func load() async {
        if loadedOnce { return }
        state = .loading
        await fetch()
    }

    public func refresh() async {
        await fetch()
    }

    private func fetch() async {
        do {
            // Access gate first — a 403 here means the viewer can't see
            // the team at all.
            let access: BusinessTeamAccessDTO = try await api.request(
                BusinessTeamEndpoints.access(businessId: businessId)
            )
            self.access = access
            guard access.hasAccess else {
                state = .error(message: "You don't have access to this team.")
                return
            }

            let memberResponse: BusinessTeamMembersResponse = try await api.request(
                BusinessTeamEndpoints.members(businessId: businessId)
            )
            members = memberResponse.members

            // Pending seats + role presets are best-effort — the screen
            // still renders the roster if either is unavailable.
            pendingSeats = await (try? fetchPendingSeats()) ?? pendingSeats
            if rolePresets.isEmpty {
                rolePresets = await (try? fetchPresets()) ?? []
            }

            loadedOnce = true
            applyState()
        } catch {
            state = .error(
                message: (error as? APIError)?.errorDescription
                    ?? "Couldn't load your team. Try again."
            )
        }
    }

    private func fetchPendingSeats() async throws -> [BusinessSeatDTO] {
        let response: BusinessSeatsResponse = try await api.request(
            BusinessTeamEndpoints.seats(businessId: businessId)
        )
        return response.seats.filter { ($0.inviteStatus ?? "").lowercased() == "pending" }
    }

    private func fetchPresets() async throws -> [BusinessRolePresetDTO] {
        let response: BusinessRolePresetsResponse = try await api.request(
            BusinessTeamEndpoints.rolePresets(businessId: businessId)
        )
        return response.presets.sorted { $0.sortOrder < $1.sortOrder }
    }

    // MARK: - Gating

    private var canManage: Bool {
        guard let access else { return false }
        return access.isOwner == true || access.permissions.contains("team.manage")
    }

    private var canInvite: Bool {
        guard let access else { return false }
        return access.isOwner == true || access.permissions.contains("team.invite")
    }

    // MARK: - Mutations

    /// Fold a freshly-created seat invite into the pending bucket so the
    /// user sees it without waiting for a refetch.
    public func handleInvited(_ seat: BusinessSeatDTO) {
        pendingSeats.insert(seat, at: 0)
        applyState()
    }

    /// Optimistic role change with rollback. The member's role tier is
    /// updated locally from the chosen preset and the list re-grouped.
    public func changeRole(userId: String, preset: BusinessRolePresetDTO) async {
        let previous = members
        members = members.map { member in
            guard member.user?.id == userId else { return member }
            return BusinessTeamMemberDTO(
                id: member.id,
                roleBase: preset.roleBase,
                title: member.title,
                joinedAt: member.joinedAt,
                invitedAt: member.invitedAt,
                notes: member.notes,
                user: member.user
            )
        }
        applyState()
        do {
            let _: EmptyResponse = try await api.request(
                BusinessTeamEndpoints.changeRole(
                    businessId: businessId,
                    userId: userId,
                    request: BusinessChangeRoleRequest(presetKey: preset.key)
                )
            )
        } catch {
            members = previous
            applyState()
        }
    }

    /// Optimistic member removal with rollback.
    public func remove(userId: String) async {
        let previous = members
        members.removeAll { $0.user?.id == userId }
        applyState()
        do {
            let _: EmptyResponse = try await api.request(
                BusinessTeamEndpoints.removeMember(businessId: businessId, userId: userId)
            )
        } catch {
            members = previous
            applyState()
        }
    }

    /// Optimistic cancel of a pending seat invite with rollback.
    public func cancelInvite(seatId: String) async {
        let previous = pendingSeats
        pendingSeats.removeAll { $0.id == seatId }
        applyState()
        do {
            let _: EmptyResponse = try await api.request(
                BusinessTeamEndpoints.cancelSeat(businessId: businessId, seatId: seatId)
            )
        } catch {
            pendingSeats = previous
            applyState()
        }
    }

    // MARK: - Permissions (Manage permissions sheet)

    /// Fetch the effective permission set for one member.
    public func memberPermissions(userId: String) async -> BusinessPermissionLoadResult {
        do {
            let response: BusinessMemberPermissionsResponse = try await api.request(
                BusinessTeamEndpoints.memberPermissions(businessId: businessId, userId: userId)
            )
            return .success(response.permissions)
        } catch {
            return .failure(
                BusinessPermissionsLoadError(
                    message: (error as? APIError)?.errorDescription ?? "Couldn't load permissions."
                )
            )
        }
    }

    /// Toggle one scoped permission for a member. Returns true on success.
    public func togglePermission(userId: String, permission: String, allowed: Bool) async -> Bool {
        do {
            let _: EmptyResponse = try await api.request(
                BusinessTeamEndpoints.togglePermission(
                    businessId: businessId,
                    userId: userId,
                    request: BusinessTogglePermissionRequest(permission: permission, allowed: allowed)
                )
            )
            return true
        } catch {
            return false
        }
    }

    // MARK: - State projection

    private func applyState() {
        let manage = canManage
        let invite = canInvite

        let sections = buildSections(canManage: manage)
        let pending = buildPending(canManage: manage)

        if sections.isEmpty, pending.isEmpty {
            state = .empty(canInvite: invite)
            return
        }
        state = .loaded(
            BusinessTeamContent(
                sections: sections,
                pending: pending,
                canManage: manage,
                canInvite: invite
            )
        )
    }

    private func buildSections(canManage: Bool) -> [BusinessTeamSection] {
        let grouped = Dictionary(grouping: members.filter { $0.user != nil }) {
            BusinessRole.parse($0.roleBase)
        }
        return grouped.keys
            .sorted { $0.rank > $1.rank }
            .compactMap { role -> BusinessTeamSection? in
                guard let bucket = grouped[role], !bucket.isEmpty else { return nil }
                let rows = bucket.map { row(forMember: $0, role: role, canManage: canManage) }
                return BusinessTeamSection(id: role.rawValue, role: role, rows: rows)
            }
    }

    private func buildPending(canManage: Bool) -> [BusinessTeamPendingRow] {
        pendingSeats.map { seat in
            let role = BusinessRole.parse(seat.roleBase)
            let name = seat.displayName?.nilIfBlank ?? seat.inviteEmail?.nilIfBlank ?? "Invited seat"
            return BusinessTeamPendingRow(
                id: seat.id,
                seatId: seat.id,
                name: name,
                email: seat.inviteEmail,
                role: role,
                invitedText: invitedText(for: seat),
                canManage: canManage
            )
        }
    }

    private func row(
        forMember member: BusinessTeamMemberDTO,
        role: BusinessRole,
        canManage: Bool
    ) -> BusinessTeamMemberRow {
        let user = member.user
        let userId = user?.id ?? member.id
        let name = Self.displayName(for: member)
        return BusinessTeamMemberRow(
            id: userId,
            userId: userId,
            name: name,
            email: user?.email,
            role: role,
            avatarGradient: BusinessTeamAvatarTone.tone(for: userId).gradient,
            joinedText: joinedText(for: member),
            canManage: canManage && role != .owner
        )
    }

    // MARK: - Helpers (pure)

    static func displayName(for member: BusinessTeamMemberDTO) -> String {
        BusinessTeamProjection.displayName(for: member)
    }

    private func joinedText(for member: BusinessTeamMemberDTO) -> String? {
        let raw = member.joinedAt ?? member.invitedAt
        guard let relative = Self.formatRelativeTime(
            raw,
            now: now(),
            calendar: calendar,
            timeZone: timeZone
        ) else {
            return nil
        }
        return "Joined \(relative)"
    }

    private func invitedText(for seat: BusinessSeatDTO) -> String? {
        guard let relative = Self.formatRelativeTime(
            seat.createdAt,
            now: now(),
            calendar: calendar,
            timeZone: timeZone
        ) else {
            return nil
        }
        return "Invited \(relative)"
    }

    public static func formatRelativeTime(
        _ raw: String?,
        now: Date,
        calendar: Calendar,
        timeZone: TimeZone
    ) -> String? {
        BusinessTeamDateFormatter.formatRelativeTime(
            raw,
            now: now,
            calendar: calendar,
            timeZone: timeZone
        )
    }
}

public struct BusinessPermissionsLoadError: Error, Sendable, Equatable {
    public let message: String
}

public typealias BusinessPermissionLoadResult = Result<[String], BusinessPermissionsLoadError>

private enum BusinessTeamProjection {
    static func displayName(for member: BusinessTeamMemberDTO) -> String {
        if let name = member.user?.name?.nilIfBlank { return name }
        if let username = member.user?.username?.nilIfBlank { return "@\(username)" }
        if let email = member.user?.email?.nilIfBlank { return email }
        return "Team member"
    }
}

private enum BusinessTeamDateFormatter {
    static func formatRelativeTime(
        _ raw: String?,
        now: Date,
        calendar: Calendar,
        timeZone: TimeZone
    ) -> String? {
        guard let date = parseDate(raw) else { return nil }
        let interval = now.timeIntervalSince(date)
        if interval < 60 { return "just now" }
        if interval < 3600 { return "\(Int(interval / 60))m ago" }
        if interval < 86400 { return "\(Int(interval / 3600))h ago" }
        var cal = calendar
        cal.timeZone = timeZone
        let startOfNow = cal.startOfDay(for: now)
        let startOfDate = cal.startOfDay(for: date)
        let dayDelta = cal.dateComponents([.day], from: startOfDate, to: startOfNow).day ?? 0
        if dayDelta == 1 { return "yesterday" }
        if dayDelta < 7 { return "\(dayDelta)d ago" }
        if dayDelta < 30 { return "\(dayDelta / 7)w ago" }
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = timeZone
        formatter.dateFormat = "MMM d"
        return formatter.string(from: date)
    }

    private static func parseDate(_ raw: String?) -> Date? {
        guard let raw, !raw.isEmpty else { return nil }
        let iso8601 = ISO8601DateFormatter()
        iso8601.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let iso8601NoFraction = ISO8601DateFormatter()
        iso8601NoFraction.formatOptions = [.withInternetDateTime]
        return iso8601.date(from: raw) ?? iso8601NoFraction.date(from: raw)
    }
}

private extension String {
    var nilIfBlank: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : self
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/businesses/team/BusinessTeamViewModel.kt
@file:Suppress("PackageNaming", "TooManyFunctions", "LongMethod", "CyclomaticComplexMethod")

package app.pantopus.android.ui.screens.businesses.team

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.businesses.BusinessAccessDto
import app.pantopus.android.data.api.models.businesses.BusinessRolePresetDto
import app.pantopus.android.data.api.models.businesses.BusinessSeatDto
import app.pantopus.android.data.api.models.businesses.BusinessTeamMemberDto
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.businesses.BusinessTeamRepository
import app.pantopus.android.data.network.NetworkMonitor
import app.pantopus.android.ui.screens.shared.list_of_rows.GradientPair
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale
import javax.inject.Inject

/** Nav arg key for the business id consumed via [SavedStateHandle]. */
const val BUSINESS_TEAM_ID_KEY = "businessId"

// ─── Projected view models (parity with iOS) ──────────────────────────

/** One member row in a role section. */
data class BusinessTeamMemberRow(
    val id: String,
    val userId: String,
    val name: String,
    val email: String?,
    val role: BusinessRole,
    val avatarGradient: GradientPair,
    val joinedText: String?,
    val canManage: Boolean,
)

/** One role group section (owner → viewer order). */
data class BusinessTeamSection(
    val id: String,
    val role: BusinessRole,
    val rows: List<BusinessTeamMemberRow>,
) {
    val headerTitle: String get() = role.pluralLabel
}

/** One pending seat-invite row. */
data class BusinessTeamPendingRow(
    val id: String,
    val seatId: String,
    val name: String,
    val email: String?,
    val role: BusinessRole,
    val invitedText: String?,
    val canManage: Boolean,
)

/** Loaded payload for the Team screen. */
data class BusinessTeamContent(
    val sections: List<BusinessTeamSection>,
    val pending: List<BusinessTeamPendingRow>,
    val canManage: Boolean,
    val canInvite: Boolean,
)

/** Render state for the Team screen. */
sealed interface BusinessTeamUiState {
    data object Loading : BusinessTeamUiState

    data class Loaded(val content: BusinessTeamContent) : BusinessTeamUiState

    data class Empty(val canInvite: Boolean) : BusinessTeamUiState

    data class Error(val message: String) : BusinessTeamUiState
}

/**
 * Owner-side business team & roles roster. Mirrors iOS
 * `BusinessTeamViewModel` — reads the businessIam members list (grouped by
 * `role_base`) plus the businessSeats pending invites, and gates actions on
 * `GET /:id/me`.
 */
@HiltViewModel
class BusinessTeamViewModel
    @Inject
    constructor(
        private val repo: BusinessTeamRepository,
        networkMonitor: NetworkMonitor,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        val businessId: String = savedStateHandle[BUSINESS_TEAM_ID_KEY] ?: ""

        /** Connectivity for the offline banner. */
        val isOnline: StateFlow<Boolean> = networkMonitor.isOnline

        private val _state = MutableStateFlow<BusinessTeamUiState>(BusinessTeamUiState.Loading)
        val state: StateFlow<BusinessTeamUiState> = _state.asStateFlow()

        private val _rolePresets = MutableStateFlow<List<BusinessRolePresetDto>>(emptyList())
        val rolePresets: StateFlow<List<BusinessRolePresetDto>> = _rolePresets.asStateFlow()

        private var access: BusinessAccessDto? = null
        private var members: List<BusinessTeamMemberDto> = emptyList()
        private var pendingSeats: List<BusinessSeatDto> = emptyList()
        private var loadedOnce = false

        /** Idempotent — re-running won't refetch once content is loaded. */
        fun load() {
            if (loadedOnce) return
            reload()
        }

        /** Pull-to-refresh / retry. */
        fun refresh() = reload()

        private fun reload() {
            _state.value = BusinessTeamUiState.Loading
            viewModelScope.launch {
                when (val accessResult = repo.access(businessId)) {
                    is NetworkResult.Failure -> {
                        _state.value = BusinessTeamUiState.Error(accessResult.error.message)
                        return@launch
                    }
                    is NetworkResult.Success -> {
                        access = accessResult.data
                        if (!accessResult.data.hasAccess) {
                            _state.value = BusinessTeamUiState.Error("You don't have access to this team.")
                            return@launch
                        }
                    }
                }

                when (val memberResult = repo.members(businessId)) {
                    is NetworkResult.Failure -> {
                        _state.value = BusinessTeamUiState.Error(memberResult.error.message)
                        return@launch
                    }
                    is NetworkResult.Success -> {
                        members = memberResult.data.members
                    }
                }

                // Pending seats + presets are best-effort.
                (repo.seats(businessId) as? NetworkResult.Success)?.let { result ->
                    pendingSeats = result.data.seats.filter { (it.inviteStatus ?: "").lowercase() == "pending" }
                }
                if (_rolePresets.value.isEmpty()) {
                    (repo.rolePresets(businessId) as? NetworkResult.Success)?.let { result ->
                        _rolePresets.value = result.data.presets.sortedBy { it.sortOrder }
                    }
                }

                loadedOnce = true
                applyState()
            }
        }

        // ─── Gating ───────────────────────────────────────────────

        private fun canManage(): Boolean {
            val a = access ?: return false
            return a.isOwner || a.permissions.contains("team.manage")
        }

        private fun canInvite(): Boolean {
            val a = access ?: return false
            return a.isOwner || a.permissions.contains("team.invite")
        }

        // ─── Mutations ────────────────────────────────────────────

        /** Fold a freshly-created seat invite into the pending bucket. */
        fun handleInvited(seat: BusinessSeatDto) {
            pendingSeats = listOf(seat) + pendingSeats
            applyState()
        }

        /** Optimistic role change with rollback. */
        fun changeRole(
            userId: String,
            preset: BusinessRolePresetDto,
        ) {
            val previous = members
            members =
                members.map { member ->
                    if (member.user?.id == userId) member.copy(roleBase = preset.roleBase) else member
                }
            applyState()
            viewModelScope.launch {
                when (repo.changeRole(businessId, userId, preset.key)) {
                    is NetworkResult.Success -> Unit
                    is NetworkResult.Failure -> {
                        members = previous
                        applyState()
                    }
                }
            }
        }

        /** Optimistic member removal with rollback. */
        fun remove(userId: String) {
            val previous = members
            members = previous.filterNot { it.user?.id == userId }
            applyState()
            viewModelScope.launch {
                when (repo.removeMember(businessId, userId)) {
                    is NetworkResult.Success -> Unit
                    is NetworkResult.Failure -> {
                        members = previous
                        applyState()
                    }
                }
            }
        }

        /** Optimistic cancel of a pending seat invite with rollback. */
        fun cancelInvite(seatId: String) {
            val previous = pendingSeats
            pendingSeats = previous.filterNot { it.id == seatId }
            applyState()
            viewModelScope.launch {
                when (repo.cancelSeat(businessId, seatId)) {
                    is NetworkResult.Success -> Unit
                    is NetworkResult.Failure -> {
                        pendingSeats = previous
                        applyState()
                    }
                }
            }
        }

        /** Fetch the effective permission set for one member. */
        suspend fun memberPermissions(userId: String): Result<List<String>> =
            when (val result = repo.memberPermissions(businessId, userId)) {
                is NetworkResult.Success -> Result.success(result.data.permissions)
                is NetworkResult.Failure -> Result.failure(result.error)
            }

        /** Toggle one scoped permission for a member. Returns true on success. */
        suspend fun togglePermission(
            userId: String,
            permission: String,
            allowed: Boolean,
        ): Boolean = repo.togglePermission(businessId, userId, permission, allowed) is NetworkResult.Success

        // ─── State projection ─────────────────────────────────────

        private fun applyState() {
            val manage = canManage()
            val invite = canInvite()
            val now = Instant.now()
            val zone = ZoneId.systemDefault()

            val sections =
                members
                    .filter { it.user != null }
                    .groupBy { BusinessRole.parse(it.roleBase) }
                    .toList()
                    .sortedByDescending { it.first.rank }
                    .mapNotNull { (role, bucket) ->
                        if (bucket.isEmpty()) {
                            null
                        } else {
                            BusinessTeamSection(
                                id = role.wire,
                                role = role,
                                rows = bucket.map { rowForMember(it, role, manage, now, zone) },
                            )
                        }
                    }

            val pending =
                pendingSeats.map { seat ->
                    val role = BusinessRole.parse(seat.roleBase)
                    BusinessTeamPendingRow(
                        id = seat.id,
                        seatId = seat.id,
                        name = seat.displayName?.ifBlank { null } ?: seat.inviteEmail?.ifBlank { null } ?: "Invited seat",
                        email = seat.inviteEmail,
                        role = role,
                        invitedText = relativeText(seat.createdAt, now, zone)?.let { "Invited $it" },
                        canManage = manage,
                    )
                }

            _state.value =
                if (sections.isEmpty() && pending.isEmpty()) {
                    BusinessTeamUiState.Empty(canInvite = invite)
                } else {
                    BusinessTeamUiState.Loaded(
                        BusinessTeamContent(sections = sections, pending = pending, canManage = manage, canInvite = invite),
                    )
                }
        }

        private fun rowForMember(
            member: BusinessTeamMemberDto,
            role: BusinessRole,
            canManage: Boolean,
            now: Instant,
            zone: ZoneId,
        ): BusinessTeamMemberRow {
            val userId = member.user?.id ?: member.id
            val name = displayName(member)
            return BusinessTeamMemberRow(
                id = userId,
                userId = userId,
                name = name,
                email = member.user?.email,
                role = role,
                avatarGradient = BusinessTeamAvatarTone.toneFor(userId).gradient,
                joinedText = relativeText(member.joinedAt ?: member.invitedAt, now, zone)?.let { "Joined $it" },
                canManage = canManage && role != BusinessRole.Owner,
            )
        }

        companion object {
            private const val SECONDS_PER_MINUTE = 60L
            private const val SECONDS_PER_HOUR = 3_600L
            private const val SECONDS_PER_DAY = 86_400L
            private const val ONE_DAY = 1L
            private const val DAYS_IN_WEEK = 7L
            private const val DAYS_IN_MONTH = 30L

            fun displayName(member: BusinessTeamMemberDto): String {
                member.user?.name?.takeIf { it.isNotBlank() }?.let { return it }
                member.user?.username?.takeIf { it.isNotBlank() }?.let { return "@$it" }
                member.user?.email?.takeIf { it.isNotBlank() }?.let { return it }
                return "Team member"
            }

            private val iso8601: DateTimeFormatter = DateTimeFormatter.ISO_DATE_TIME

            private fun parseInstant(raw: String?): Instant? {
                if (raw.isNullOrEmpty()) return null
                return runCatching { Instant.parse(raw) }
                    .getOrElse { runCatching { iso8601.parse(raw, Instant::from) }.getOrNull() }
            }

            internal fun relativeText(
                raw: String?,
                now: Instant,
                zone: ZoneId,
            ): String? {
                val instant = parseInstant(raw) ?: return null
                val interval = now.epochSecond - instant.epochSecond
                val startOfNow = now.atZone(zone).toLocalDate()
                val startOfDate = instant.atZone(zone).toLocalDate()
                val dayDelta = ChronoUnit.DAYS.between(startOfDate, startOfNow)
                val formatter = DateTimeFormatter.ofPattern("MMM d", Locale.US).withZone(zone)
                return when {
                    interval < SECONDS_PER_MINUTE -> "just now"
                    interval < SECONDS_PER_HOUR -> "${interval / SECONDS_PER_MINUTE}m ago"
                    interval < SECONDS_PER_DAY -> "${interval / SECONDS_PER_HOUR}h ago"
                    dayDelta == ONE_DAY -> "yesterday"
                    dayDelta < DAYS_IN_WEEK -> "${dayDelta}d ago"
                    dayDelta < DAYS_IN_MONTH -> "${dayDelta / DAYS_IN_WEEK}w ago"
                    else -> formatter.format(instant)
                }
            }
        }
    }
